File: leetcode/maximum_subarray.cpp
// class Solution {
// public:
//     int maxSubArray(vector<int>& nums) {
        
//         const int size = nums.size();
//         int curSum = nums[0];
//         int ans = nums[0];
        
//         for (int it = 1; it < size; it++){
//             curSum = max(nums[it], curSum+nums[it]);
//             ans = max(curSum,ans);
//             }

//         return ans;
//     }
    
// };
// constantly update the maximum value.

class Solution {
public:
    int maxSubArray(vector<int>& nums) {
        
        const int size = nums.size();
        int curSum = 0;
        int ans = nums[0];
        
        for (int it = 0; it < size; it++){
            if (curSum < 0){
                curSum = 0;
            }
            
            curSum += nums[it];
            ans = max(curSum,ans);
        }
        
        return ans;
    }
};
// realize that we don't need to add a negative value.

File: leetcode/maximum_subarray.py
# class Solution:
#     def maxSubArray(self, nums: List[int]) -> int:
#         length = len(nums) 
#         sum_array = [nums[0]]
#         for i in range(1, length):
#             sum_array.append(max(nums[i] + sum_array[i-1], nums[i]))
#         return max(sum_array)
# Making extra array storing maximum

class Solution:
    def maxSubArray(self, nums: List[int]) -> int:
        length = len(nums)
        cur_max = 0
        ans = nums[0]
        for i in range(0, length):
            if (cur_max < 0):
                cur_max = 0
            cur_max += nums[i]
            ans = max(cur_max, ans)
        return ans
# Constantly update the maximum, finding out that we don't need to add 'negative value'.
